File: coinChange.py
class CoinChange:
    def brute_force(self, coins, amount):
        if amount == 0:
            return 0
        if amount < 0:
            return -1
        res = (amount + 1)
        for coin in coins:
            sub_res = self.brute_force(coins, amount - coin)
            if sub_res >= 0:
                res = min(res, 1 + sub_res)
        return res if res != (amount + 1) else -1

    def memoization(self, coins, amount):
        memo = {0 : 0}
        def dp(a):
            if a in memo:
                return memo[a]
            if a < 0:
                return a
            res = (a + 1)
            for coin in coins:
                sub_res = dp(a - coin)
                if sub_res >= 0:
                    res = min(res, 1 + sub_res)
            memo[a] = res if res != (a + 1) else -1
            return memo[a]
        return dp(amount)

def get_input():
    coins = input("Enter coin denominations (space-separated): ")
    coins = [int(c) for c in coins.strip().split()]
    while True:
        amount_input = input("Enter amount to make change for: ")
        if amount_input.isdigit():
            amount = int(amount_input)
            break
        print("Invalid input. Please enter a valid integer.")
    return coins, amount

def main():
    coin_change = CoinChange()
    while True:
        method = input("Choose method ('brute', 'memo' or 'exit'): ").strip().lower()
        if method == 'exit':
            break
        if method not in ['brute', 'memo']:
            print("Invalid choice. Please choose 'brute', 'memo' or 'exit'.")
            continue
        coins, amount = get_input()
        if method == 'brute':
            result = coin_change.brute_force(coins, amount)
        else:
            result = coin_change.memoization(coins, amount)
        if result == -1:
            print(f"Cannot make change for {amount} with given coins.")
        else:
            print(f"Minimum coins needed: {result}")

if __name__ == "__main__":
    main()

File: coinChange.ts
import process from "node:process";
import { createInterface } from "node:readline";

class CoinChange {
  bruteForce(coins: number[], amount: number): number {
    if (amount === 0) {
      return 0;
    }
    if (amount < 0) {
      return -1;
    }
    let res: number = amount + 1;
    for (const coin of coins) {
      const subRes: number = this.bruteForce(coins, amount - coin);
      if (subRes >= 0) {
        res = Math.min(res, 1 + subRes);
      }
    }
    return res !== (amount + 1) ? res : -1;
  }

  memoization(coins: number[], amount: number): number {
    const memo: Record<number, number> = { 0: 0 };

    const dp = (a: number): number => {
      if (a in memo) {
        return memo[a];
      }
      if (a < 0) {
        return a;
      }
      let res: number = a + 1;
      for (const coin of coins) {
        const subRes: number = dp(a - coin);
        if (subRes >= 0) {
          res = Math.min(res, 1 + subRes);
        }
      }
      memo[a] = res !== (a + 1) ? res : -1;
      return memo[a];
    };

    return dp(amount);
  }
}

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

function question(prompt: string): Promise<string> {
  return new Promise((resolve) => {
    rl.question(prompt, (answer: string) => {
      resolve(answer);
    });
  });
}

async function getInput(): Promise<[number[], number]> {
  const coinsInput: string = await question(
    "Enter coin denominations (space-separated): ",
  );
  const coins: number[] = coinsInput.split(" ").map((c) => parseInt(c, 10));

  let amount: number = 0;
  while (true) {
    const amountInput: string = await question(
      "Enter amount to make change for: ",
    );
    if (/^\d+$/.test(amountInput)) {
      amount = parseInt(amountInput, 10);
      break;
    } else {
      console.log("Invalid input. Please enter a valid integer.");
    }
  }

  return [coins, amount];
}

async function main(): Promise<void> {
  const coinChange: CoinChange = new CoinChange();

  const processMethod = async (): Promise<void> => {
    const method: string = await question(
      "Choose method ('brute', 'memo' or 'exit'): ",
    );
    const methodLower: string = method.trim().toLowerCase();

    if (methodLower === "exit") {
      rl.close();
      return;
    }

    if (methodLower !== "brute" && methodLower !== "memo") {
      console.log("Invalid choice. Please choose 'brute', 'memo' or 'exit'.");
      return processMethod();
    }

    try {
      const [coins, amount]: [number[], number] = await getInput();

      let result: number;
      if (methodLower === "brute") {
        result = coinChange.bruteForce(coins, amount);
      } else {
        result = coinChange.memoization(coins, amount);
      }

      if (result === -1) {
        console.log(`Cannot make change for ${amount} with given coins.`);
      } else {
        console.log(`Minimum coins needed: ${result}`);
      }
    } catch (error: unknown) {
      console.error("An error occurred:", error);
    }

    return processMethod();
  };

  await processMethod();
}

main();
